### PR TITLE
cleanup: fix AWS role loading

### PR DESCRIPTION
**Pull request by nrainer-materialize**

### CODE REVIEW

File: ci/cleanup/aws.py [hunk 1]
```diff
@@ -125,6 +123,28 @@ def clean_up_iam() -> None:
...
 
+def get_testdrive_roles(client: Any) -> list[Any]:
+    roles = []
+    marker = None
+
+    while True:
+        if marker is not None:
+            response = client.list_roles(MaxItems=1000, Marker=marker)
+        else:
+            response = client.list_roles(MaxItems=1000)
```
Comment (def-) [inline ci/cleanup/aws.py:134]: Why not just this?
```suggestion
        response = client.list_roles(MaxItems=1000, Marker=marker)
```

File: ci/cleanup/aws.py [hunk 2]
```diff
@@ -125,6 +123,28 @@ def clean_up_iam() -> None:
...
+        roles.extend(response.get("Roles", []))
+        is_truncated = response.get("IsTruncated")
+
+        if is_truncated:
+            marker = response.get("Marker")
+        else:
+            break
+
+    roles = [r for r in roles if r["RoleName"].startswith("testdrive")]
+    return roles
```
Comment (def-) [inline ci/cleanup/aws.py:145]:
```suggestion
    return [r for r in roles if r["RoleName"].startswith("testdrive")]
```

File: ci/cleanup/aws.py [hunk 3]
```diff
@@ -125,6 +123,28 @@ def clean_up_iam() -> None:
             print(f"Skipping role {role_name}")
 
 
+def get_testdrive_roles(client: Any) -> list[Any]:
+    roles = []
+    marker = None
+
+    while True:
```
Comment (def-) [inline ci/cleanup/aws.py:130]: I guess not too much danger of an endless loop, timeout is 10 min anyway

**Review by def- — APPROVED** [review]
Just minor suggestions

File: ci/cleanup/aws.py (hunk 1 above)
Comment (nrainer-materialize) [reply ci/cleanup/aws.py:134]: It did not work due to:

```
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter Marker, value: None, type: <class 'NoneType'>, valid types: <class 'str'>
```

Comment (nrainer-materialize) [reply ci/cleanup/aws.py:134]: And an empty parameter value neither did:
```
botocore.exceptions.ParamValidationError: Parameter validation failed: Invalid length for parameter Marker, value: 0, valid min length: 1
```

**Review by benesch — COMMENTED** [review]
Aha, nice find! Approach LGTM but deferring approval of the code to Dennis.

File: ci/cleanup/aws.py (hunk 3 above)
Comment (nrainer-materialize) [reply ci/cleanup/aws.py:130]: Yeah. There is a break statement, which is reached when `is_truncated` is `False`.

**Comment by benesch** [comment]
Consider also using the built-in boto3 paginator, which automatically handles making paginated API calls: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html

**Comment by nrainer-materialize** [comment]
> Consider also using the built-in boto3 paginator, which automatically handles making paginated API calls: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html

Yeah, that's cleaner. Done.

**Pull request merged**